src/blockMonitor/blockCache.ts

```diff
@@ -215,20 +189,66 @@ export class BlockCache implements ReadOnlyBlockCache {
         return this.blockStubsByHash.has(blockHash);
     }
 
+    /**
+     * Iterator over all the blocks in the ancestry of the block with hash `initialBlockHash` (inclusive).
+     * @param initialBlockHash
+     */
+    private *ancestry(initialBlockHash: string): IterableIterator<IBlockStub> {
+        let curBlock = this.getBlockStub(initialBlockHash);
+        while (curBlock !== null) {
+            yield curBlock;
+            curBlock = this.getBlockStub(curBlock.parentHash);
+        }
+    }
+
+    /**
+     * Finds and returns the nearest ancestor that satisfies `predicate`.
+     * Returns `null` if no such ancestor is found.
+     */
+    public findAncestor(initialBlockHash: string, predicate: (block: IBlockStub) => boolean): IBlockStub | null {
+        for (const block of this.ancestry(initialBlockHash)) {
+            if (predicate(block)) {
+                return block;
+            }
+        }
+        return null;
+    }
+
+    /**
+     * Returns the IBlockStub of the oldest ancestor of `blockStub` that is stored in the blockCache.
+     * @throws ArgumentError if `blockHash` is not in the blockCache.
+     * @param blockHash
+     */
+    public getOldestAncestorInCache(blockHash: string): IBlockStub {
+        if (!this.hasBlock(blockHash)) {
+            throw new ArgumentError(`The block with hash ${blockHash} is not in cache`);
+        }
+
+        // Find the deepest ancestor that is in cache
+        const result = this.findAncestor(blockHash, block => !this.hasBlock(block.parentHash));
+
+        if (!result) {
+            // This should never happen
+            throw new ApplicationError(
+                `An error occurred while searching the oldest block in cache after a catastrophic reorg. This is a bug.`
```

Comment:
Simplified message to `An error occurred while searching the oldest ancestor in cache. This is a bug.`.

Reply:
KK, you may actually find you dont need this method anymore at the end of this code review. But if you do, I would try to still be more explicit. Something like:

`No ancestor block found that was *not* in the cache.`

I wouldn't include a message like `this is a bug` in errors. An application error will hopefully be logged in the errors file, and all of these should be investigated for bugs.

Reply:
Well, the error condition is that "findAncestor" returns null, which cannot happen because the block itself already satisfied the condition (and we already checked that is in the cache).

Changed message to `An error occurred while searching for the oldest ancestor in cache`, and added a comment explaining why it cannot happen. f1913c4